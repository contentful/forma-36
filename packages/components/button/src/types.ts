import { ComponentVariant } from '@contentful/f36-core';

export type ButtonSize = 'small' | 'medium' | 'large';

export type ButtonVariant =
  | Exclude<ComponentVariant, 'warning'>
  | 'transparent';

export type ButtonStylesProps = {
  variant: ButtonVariant;
  size: ButtonSize;
  isActive: boolean;
  isDisabled: boolean;
  isFullWidth: boolean;
};
