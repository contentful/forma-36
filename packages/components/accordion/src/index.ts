export { Accordion } from './Accordion';
export type { AccordionProps } from './Accordion';
export { AccordionHeader } from './AccordionHeader/AccordionHeader';
export type { AccordionHeaderProps } from './AccordionHeader/AccordionHeader';
export { AccordionItem } from './AccordionItem/AccordionItem';
export type { AccordionItemProps } from './AccordionItem/AccordionItem';
export { AccordionPanel } from './AccordionPanel/AccordionPanel';
export type { AccordionPanelProps } from './AccordionPanel/AccordionPanel';
