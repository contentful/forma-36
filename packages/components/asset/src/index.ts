export { Asset } from './Asset';
export type { AssetProps } from './Asset';
