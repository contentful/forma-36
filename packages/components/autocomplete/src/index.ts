export { Autocomplete } from './Autocomplete';
export type { AutocompleteProps } from './Autocomplete';
