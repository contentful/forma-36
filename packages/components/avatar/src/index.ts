export { Avatar } from './Avatar';
export type { AvatarProps } from './Avatar';
export { AvatarGroup } from './AvatarGroup';
export type { AvatarGroupProps } from './AvatarGroup';
