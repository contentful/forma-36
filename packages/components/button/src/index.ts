export { ButtonGroup } from './ButtonGroup';
export type { ButtonGroupProps } from './ButtonGroup';
export { Button } from './Button';
export type { ButtonProps } from './Button';
