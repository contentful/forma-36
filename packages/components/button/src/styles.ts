import { css } from 'emotion';
import type { CSSObject } from '@emotion/serialize';
import tokens from '@contentful/f36-tokens';
import { ButtonSize, ButtonVariant, ButtonStylesProps } from './types';

const variantActiveStyles = (variant: ButtonVariant): CSSObject => {
  switch (variant) {
    case 'primary':
      return {
        ', &:hover': {
          backgroundColor: tokens.blue700,
          borderColor: tokens.blue700,
        },
      };
    case 'secondary':
      return {
        ', &:hover': {
          backgroundColor: tokens.gray200,
        },
      };
    case 'positive':
      return {
        ', &:hover': {
          backgroundColor: tokens.green700,
          borderColor: tokens.green700,
        },
      };
    case 'negative':
      return {
        ', &:hover': {
          backgroundColor: tokens.red800,
          borderColor: tokens.red800,
        },
      };
    case 'transparent':
      return {
        ', &:hover': {
          backgroundColor: tokens.gray100,
          borderColor: tokens.gray100,
        },
      };
    default:
      return {};
  }
};

const variantToStyles = (variant: ButtonVariant): CSSObject => {
  switch (variant) {
    case 'primary':
      return {
        color: tokens.colorWhite,
        backgroundColor: tokens.blue500,
        borderColor: tokens.blue500,
        '&:hover': {
          backgroundColor: tokens.blue600,
          borderColor: tokens.blue600,
        },
        '&:active': variantActiveStyles(variant),
        '&:focus': {
          borderColor: tokens.blue600,
          boxShadow: tokens.glowPrimary,
        },
      };
    case 'secondary':
      return {
        color: tokens.gray900,
        backgroundColor: tokens.colorWhite,
        borderColor: tokens.gray300,
        '&:hover': {
          backgroundColor: tokens.gray100,
        },
        '&:active': variantActiveStyles(variant),
        '&:focus': {
          boxShadow: tokens.glowMuted,
        },
      };
    case 'positive':
      return {
        color: tokens.colorWhite,
        backgroundColor: tokens.colorPositive,
        borderColor: tokens.colorPositive,
        '&:hover': {
          backgroundColor: tokens.green600,
          borderColor: tokens.green600,
        },
        '&:active': variantActiveStyles(variant),
        '&:focus': {
          borderColor: tokens.green600,
          boxShadow: tokens.glowPositive,
        },
      };
    case 'negative':
      return {
        color: tokens.colorWhite,
        backgroundColor: tokens.red600,
        borderColor: tokens.red600,
        '&:hover': {
          backgroundColor: tokens.red700,
          borderColor: tokens.red700,
        },
        '&:active': variantActiveStyles(variant),
        '&:focus': {
          borderColor: tokens.red700,
          boxShadow: tokens.glowNegative,
        },
      };
    case 'transparent':
      return {
        color: tokens.gray800,
        background: `none`,
        borderColor: `transparent`,
        boxShadow: `none`,
        '&:hover': {
          backgroundColor: tokens.gray100,
          color: tokens.gray900,
        },
        '&:active': variantActiveStyles(variant),
        '&:focus': {
          backgroundColor: 'none',
          boxShadow: tokens.glowMuted,
        },
      };
    default:
      return {};
  }
};

const sizeToStyles = (size: ButtonSize): CSSObject => {
  switch (size) {
    case 'small':
      return {
        fontSize: tokens.fontSizeM,
        lineHeight: tokens.lineHeightCondensed,
        padding: `${tokens.spacing2Xs} ${tokens.spacingS}`,
        minHeight: `32px`,
      };
    case 'medium':
      return {
        fontSize: tokens.fontSizeM,
        lineHeight: tokens.lineHeightCondensed,
        padding: `${tokens.spacingXs} ${tokens.spacingM}`,
        minHeight: `40px`,
      };
    case 'large':
      return {
        fontSize: tokens.fontSizeXl,
        lineHeight: tokens.lineHeightXl,
        padding: `${tokens.spacingXs} ${tokens.spacingM}`,
        minHeight: `48px`,
      };
    default:
      return {};
  }
};

export const styles = {
  button: ({
    variant,
    size,
    isActive,
    isDisabled,
    isFullWidth,
  }: ButtonStylesProps) =>
    css({
      boxSizing: 'border-box',
      border: `1px solid`,
      boxShadow: '0px 1px 0px rgb(25, 37, 50, 0.08)',
      borderRadius: tokens.borderRadiusMedium,
      cursor: isDisabled ? 'not-allowed' : 'pointer',
      opacity: isDisabled ? 0.5 : 1,
      display: isFullWidth ? 'flex' : 'inline-flex',
      minWidth: isFullWidth ? '100%' : 'auto',
      justifyContent: 'center',
      alignItems: 'center',
      fontWeight: tokens.fontWeightMedium,
      outline: 'none',
      textDecoration: 'none',
      transition: `background-color ${tokens.transitionDurationDefault} ${tokens.transitionEasingDefault}`,
      ...variantToStyles(variant),
      ...sizeToStyles(size),
      ...(isActive ? variantActiveStyles(variant) : {}),
    }),
  buttonText: css({
    display: 'inline-block',
  }),
  buttonIcon: css({
    fill: 'currentColor',
  }),
  dropdownIcon: css({
    marginLeft: tokens.spacingXs,
    fill: 'currentColor',
  }),
};
