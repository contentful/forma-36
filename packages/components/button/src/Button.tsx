import React, { ElementType, HTMLProps } from 'react';
import { cx } from 'emotion';
import { Flex, CommonProps } from '@contentful/f36-core';
import { ChevronDown } from '@contentful/f36-icons';
import { Icon, IconComponent } from '@contentful/f36-icon';
import { Spinner } from '@contentful/f36-spinner';

import type { ButtonVariant, ButtonSize } from './types';
import { getStyles } from './styles';

export interface ButtonProps
  extends Omit<HTMLProps<HTMLButtonElement & HTMLAnchorElement>, 'size'>,
    CommonProps {
  children?: React.ReactNode;
  /**
   * Determines style variation of Button component
   * @default secondary
   */
  variant?: ButtonVariant;
  /**
   * Determines size variation of Button component
   * @default medium
   */
  size?: ButtonSize;
  /**
   * Applies active styles
   * @default false
   */
  isActive?: boolean;
  /**
   * Disabled interaction and applies disabled styles
   * @default false
   */
  isDisabled?: boolean;
  /**
   * Button html type attribute
   */
  type?: 'submit' | 'button' | 'reset';
  /**
   * Adds dropdown indicator icon
   */
  isDropdown?: boolean;
  /**
   * Expects any of the icon components
   */
  icon?: IconComponent;
  /**
   * Adds loading indicator icon and disables interactions
   */
  isLoading?: boolean;
  /**
   * Forces button to take 100% of the container
   */
  isFullWidth?: boolean;
}

const _Button = (props: ButtonProps, ref) => {
  const styles = getStyles();
  const {
    children,
    className,
    testId = 'cf-ui-button',
    variant = 'secondary',
    size = 'medium',
    href,
    type = !href ? 'button' : undefined,
    icon,
    isActive,
    isDisabled,
    isDropdown,
    isLoading,
    isFullWidth,
    ...otherProps
  } = props;

  const rootClassNames = cx(
    styles.button({
      variant,
      size,
      isActive,
      isDisabled,
      isFullWidth,
    }),
    className,
  );

  const Element: ElementType = href ? 'a' : 'button';

  return (
    <Element
      ref={ref}
      type={type}
      className={rootClassNames}
      href={href}
      disabled={isDisabled}
      data-test-id={testId}
      {...otherProps}
    >
      {icon && !isLoading && (
        <Flex as="span" marginRight={children ? 'spacing2Xs' : 'none'}>
          <Icon
            className={styles.buttonIcon}
            as={icon}
            size={size === 'large' ? 'medium' : 'small'}
          />
        </Flex>
      )}
      <span className={styles.buttonText}>{children}</span>
      {isLoading && (
        <Spinner
          marginLeft={children || !isLoading ? 'spacingXs' : 'none'}
          customSize={18}
          variant={variant === 'secondary' ? 'default' : 'white'}
        />
      )}
      {isDropdown && <ChevronDown className={styles.dropdownIcon} />}
    </Element>
  );
};

/**
 * @description: Buttons communicate the action that will occur when the user clicks it
 */
export const Button = React.forwardRef(_Button);
