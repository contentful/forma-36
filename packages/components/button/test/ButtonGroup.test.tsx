import React from 'react';
import { render } from '@testing-library/react';
import { ButtonGroup, Button } from '../src';
import { ChevronDown } from '@contentful/f36-icons';

describe('ButtonGroup', function () {
  it('renders collapsed buttons', () => {
    const { container } = render(
      <ButtonGroup>
        <Button>Button</Button>
        <Button icon={ChevronDown} />
      </ButtonGroup>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders separated buttons', () => {
    const { container } = render(
      <ButtonGroup variant="spaced">
        <Button>Button</Button>
        <Button icon={ChevronDown} />
      </ButtonGroup>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders additional class name', () => {
    const { container } = render(
      <ButtonGroup className="additional-class">
        <Button>Button</Button>
        <Button icon={ChevronDown} />
      </ButtonGroup>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
