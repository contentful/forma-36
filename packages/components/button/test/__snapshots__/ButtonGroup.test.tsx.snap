// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup renders additional class name 1`] = `
.emotion-6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 {
  display: inline-block;
}

.emotion-1 {
  box-sizing: border-box;
  border: 1px solid;
  box-shadow: 0px 1px 0px rgb(25,37,50,0.08);
  border-radius: 6px;
  cursor: pointer;
  opacity: 1;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  min-width: auto;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 500;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background-color 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out;
  color: #111B2B;
  background-color: #ffffff;
  border-color: #CFD9E0;
  font-size: 0.875rem;
  line-height: 1.25;
  padding: 0.5rem 1rem;
  min-height: 40px;
  border-radius: 0 !important;
  margin-right: -1px;
  z-index: 1;
}

.emotion-1:hover {
  background-color: #F7F9FA;
}

.emotion-1:active,
.emotion-1:active:hover {
  background-color: #E7EBEE;
}

.emotion-1:focus {
  box-shadow: 0px 0px 0px 3px #E7EBEE;
}

.emotion-1:first-child {
  border-bottom-left-radius: 6px !important;
  border-top-left-radius: 6px !important;
}

.emotion-1:last-child {
  border-bottom-right-radius: 6px !important;
  border-top-right-radius: 6px !important;
  margin-right: 0;
}

.emotion-1:focus {
  z-index: 2;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 0;
}

.emotion-2 {
  display: inline-block;
  fill: #0059C8;
  height: 18px;
  width: 18px;
  display: inline-block;
  fill: #0059C8;
  height: 18px;
  width: 18px;
  fill: currentColor;
}

<div
  class="additional-class emotion-6"
  data-test-id="cf-ui-button-group"
>
  <button
    class="emotion-1"
    data-test-id="cf-ui-button"
    type="button"
  >
    <span
      class="emotion-0"
    >
      Button
    </span>
  </button>
  <button
    class="emotion-1"
    data-test-id="cf-ui-button"
    type="button"
  >
    <span
      class="emotion-3"
    >
      <svg
        class="emotion-2"
        data-test-id="cf-ui-icon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M5.29289 8.29289C5.68342 7.90237 6.31658 7.90237 6.70711 8.29289L12 13.5858L17.2929 8.29289C17.6834 7.90237 18.3166 7.90237 18.7071 8.29289C19.0976 8.68342 19.0976 9.31658 18.7071 9.70711L12.7071 15.7071C12.3166 16.0976 11.6834 16.0976 11.2929 15.7071L5.29289 9.70711C4.90237 9.31658 4.90237 8.68342 5.29289 8.29289Z"
        />
      </svg>
    </span>
    <span
      class="emotion-0"
    />
  </button>
</div>
`;

exports[`ButtonGroup renders collapsed buttons 1`] = `
.emotion-6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 {
  display: inline-block;
}

.emotion-1 {
  box-sizing: border-box;
  border: 1px solid;
  box-shadow: 0px 1px 0px rgb(25,37,50,0.08);
  border-radius: 6px;
  cursor: pointer;
  opacity: 1;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  min-width: auto;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 500;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background-color 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out;
  color: #111B2B;
  background-color: #ffffff;
  border-color: #CFD9E0;
  font-size: 0.875rem;
  line-height: 1.25;
  padding: 0.5rem 1rem;
  min-height: 40px;
  border-radius: 0 !important;
  margin-right: -1px;
  z-index: 1;
}

.emotion-1:hover {
  background-color: #F7F9FA;
}

.emotion-1:active,
.emotion-1:active:hover {
  background-color: #E7EBEE;
}

.emotion-1:focus {
  box-shadow: 0px 0px 0px 3px #E7EBEE;
}

.emotion-1:first-child {
  border-bottom-left-radius: 6px !important;
  border-top-left-radius: 6px !important;
}

.emotion-1:last-child {
  border-bottom-right-radius: 6px !important;
  border-top-right-radius: 6px !important;
  margin-right: 0;
}

.emotion-1:focus {
  z-index: 2;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 0;
}

.emotion-2 {
  display: inline-block;
  fill: #0059C8;
  height: 18px;
  width: 18px;
  display: inline-block;
  fill: #0059C8;
  height: 18px;
  width: 18px;
  fill: currentColor;
}

<div
  class="emotion-6"
  data-test-id="cf-ui-button-group"
>
  <button
    class="emotion-1"
    data-test-id="cf-ui-button"
    type="button"
  >
    <span
      class="emotion-0"
    >
      Button
    </span>
  </button>
  <button
    class="emotion-1"
    data-test-id="cf-ui-button"
    type="button"
  >
    <span
      class="emotion-3"
    >
      <svg
        class="emotion-2"
        data-test-id="cf-ui-icon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M5.29289 8.29289C5.68342 7.90237 6.31658 7.90237 6.70711 8.29289L12 13.5858L17.2929 8.29289C17.6834 7.90237 18.3166 7.90237 18.7071 8.29289C19.0976 8.68342 19.0976 9.31658 18.7071 9.70711L12.7071 15.7071C12.3166 16.0976 11.6834 16.0976 11.2929 15.7071L5.29289 9.70711C4.90237 9.31658 4.90237 8.68342 5.29289 8.29289Z"
        />
      </svg>
    </span>
    <span
      class="emotion-0"
    />
  </button>
</div>
`;

exports[`ButtonGroup renders separated buttons 1`] = `
.emotion-6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 {
  display: inline-block;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 0;
}

.emotion-2 {
  display: inline-block;
  fill: #0059C8;
  height: 18px;
  width: 18px;
  display: inline-block;
  fill: #0059C8;
  height: 18px;
  width: 18px;
  fill: currentColor;
}

.emotion-1 {
  box-sizing: border-box;
  border: 1px solid;
  box-shadow: 0px 1px 0px rgb(25,37,50,0.08);
  border-radius: 6px;
  cursor: pointer;
  opacity: 1;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  min-width: auto;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 500;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background-color 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out;
  color: #111B2B;
  background-color: #ffffff;
  border-color: #CFD9E0;
  font-size: 0.875rem;
  line-height: 1.25;
  padding: 0.5rem 1rem;
  min-height: 40px;
  margin-left: 0.75rem;
  margin-right: 0.75rem;
}

.emotion-1:hover {
  background-color: #F7F9FA;
}

.emotion-1:active,
.emotion-1:active:hover {
  background-color: #E7EBEE;
}

.emotion-1:focus {
  box-shadow: 0px 0px 0px 3px #E7EBEE;
}

.emotion-1:first-child {
  margin-left: 0;
}

.emotion-1:last-child {
  margin-right: 0;
}

<div
  class="emotion-6"
  data-test-id="cf-ui-button-group"
>
  <button
    class="emotion-1"
    data-test-id="cf-ui-button"
    type="button"
  >
    <span
      class="emotion-0"
    >
      Button
    </span>
  </button>
  <button
    class="emotion-1"
    data-test-id="cf-ui-button"
    type="button"
  >
    <span
      class="emotion-3"
    >
      <svg
        class="emotion-2"
        data-test-id="cf-ui-icon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M5.29289 8.29289C5.68342 7.90237 6.31658 7.90237 6.70711 8.29289L12 13.5858L17.2929 8.29289C17.6834 7.90237 18.3166 7.90237 18.7071 8.29289C19.0976 8.68342 19.0976 9.31658 18.7071 9.70711L12.7071 15.7071C12.3166 16.0976 11.6834 16.0976 11.2929 15.7071L5.29289 9.70711C4.90237 9.31658 4.90237 8.68342 5.29289 8.29289Z"
        />
      </svg>
    </span>
    <span
      class="emotion-0"
    />
  </button>
</div>
`;
